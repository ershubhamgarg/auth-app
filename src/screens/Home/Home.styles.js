import { StyleSheet } from "react-native";
import {
  BG_COLOR,
  BUTTON_COLOR,
  LIGHT_BLUE,
} from "../../constants/styleConstants";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: BG_COLOR,
  },
  cardContainer: { flex: 0.8, alignItems: "center", justifyContent: "center" },
  card: {
    borderWidth: 1,
    width: "80%",
    padding: 20,
    // borderRadius: 10,
    backgroundColor: BG_COLOR,
    shadowOffset: { width: 1, height: 1 },
    shadowOpacity: 0.5,
  },
  buttonContainer: { flex: 0.2, padding: 40 },
  name: {
    color: BUTTON_COLOR,
    fontWeight: "800",
    fontSize: 20,
  },
  email: {
    color: BUTTON_COLOR,
    fontWeight: "500",
    fontSize: 15,
  },
});
