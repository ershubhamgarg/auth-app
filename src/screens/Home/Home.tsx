import { Alert, StatusBar, Text, useColorScheme, View } from "react-native";
import { SafeAreaProvider, SafeAreaView } from "react-native-safe-area-context";
import Button from "../../components/Button/Button";
import { CommonActions, useNavigation } from "@react-navigation/native";
import { MainStackParamList } from "../../navigation/MainStack";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { styles } from "./Home.styles";
import { useAuth } from "../../context/AuthContext";
export const Home = () => {
  const { user, logout } = useAuth();
  const navigation =
    useNavigation<NativeStackNavigationProp<MainStackParamList>>();

  const handleLogout = () => {
    Alert.alert("Logout", "Are you sure you want to logout ?", [
      { text: "No" },
      {
        text: "Yes",
        style: "destructive",
        onPress: () => {
          logout();

          navigation.dispatch(
            CommonActions.reset({
              index: 0,
              routes: [{ name: "Login" }],
            })
          );
        },
      },
    ]);
  };

  return (
    <SafeAreaProvider>
      <SafeAreaView style={styles.container}>
        <View style={styles.cardContainer}>
          <View style={styles.card}>
            <Text style={styles.name}>{`Name : ${user?.name}`}</Text>
            <Text style={styles.email}>{`E-mail : ${user?.email}`}</Text>
          </View>
        </View>
        <View style={styles.buttonContainer}>
          <Button label="Log out" onPress={() => handleLogout()} />
        </View>
      </SafeAreaView>
    </SafeAreaProvider>
  );
};
