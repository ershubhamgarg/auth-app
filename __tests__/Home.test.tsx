import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { Alert } from "react-native";
import { Home } from "../src/screens/Home";

// --- Mock navigation ---
const mockDispatch = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actual = jest.requireActual("@react-navigation/native");
  return {
    ...actual,
    useNavigation: () => ({
      dispatch: mockDispatch,
    }),
    CommonActions: {
      reset: jest.fn((payload) => payload), // return payload so we can inspect it
    },
  };
});

// --- Mock Auth Context ---
const mockLogout = jest.fn();
const mockUser = { name: "John Doe", email: "john@example.com" };

jest.mock("../src/context/AuthContext", () => ({
  useAuth: () => ({
    user: mockUser,
    logout: mockLogout,
  }),
}));

// --- Mock Alert.alert ---
jest.spyOn(Alert, "alert").mockImplementation(jest.fn());

describe("Home Screen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test("matches snapshot", () => {
    const { toJSON } = render(<Home />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("renders user name and email", async () => {
    const { getByTestId, getByText } = render(<Home />);
    expect(getByTestId("card-container")).toBeTruthy();
    expect(getByText("Name : John Doe")).toBeTruthy();
    expect(getByText("E-mail : john@example.com")).toBeTruthy();
  });

  test("shows logout confirmation alert when pressing logout", () => {
    const { getByText } = render(<Home />);
    const logoutButton = getByText("Log out");
    fireEvent.press(logoutButton);

    expect(Alert.alert).toHaveBeenCalledWith(
      "Logout",
      "Are you sure you want to logout ?",
      expect.any(Array)
    );
  });

  test("calls logout and navigates to Login when confirming logout", () => {
    const alertSpy = jest
      .spyOn(Alert, "alert")
      .mockImplementation((title, msg, buttons) => {
        // Simulate pressing "Yes"
        const yesButton = buttons?.find((b) => b.text === "Yes");
        yesButton?.onPress?.();
      });

    const { getByText } = render(<Home />);
    fireEvent.press(getByText("Log out"));

    expect(mockLogout).toHaveBeenCalled();

    // Verify navigation reset call
    expect(mockDispatch).toHaveBeenCalledWith({
      index: 0,
      routes: [{ name: "Login" }],
    });

    alertSpy.mockRestore();
  });

  test('does not logout when pressing "No" on alert', () => {
    const alertSpy = jest
      .spyOn(Alert, "alert")
      .mockImplementation((title, msg, buttons) => {
        const noButton = buttons?.find((b) => b.text === "No");
        noButton?.onPress?.(); // simulate pressing "No"
      });

    const { getByText } = render(<Home />);
    fireEvent.press(getByText("Log out"));

    expect(mockLogout).not.toHaveBeenCalled();
    expect(mockDispatch).not.toHaveBeenCalled();

    alertSpy.mockRestore();
  });
});
