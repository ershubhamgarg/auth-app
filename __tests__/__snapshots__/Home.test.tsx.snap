// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home Screen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 0.8,
        "justifyContent": "center",
      }
    }
    testID="card-container"
  >
    <View
      style={
        {
          "backgroundColor": undefined,
          "borderWidth": 1,
          "padding": 20,
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.5,
          "width": "80%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#132440",
            "fontSize": 20,
            "fontWeight": "800",
          }
        }
      >
        Name : John Doe
      </Text>
      <Text
        style={
          {
            "color": "#132440",
            "fontSize": 15,
            "fontWeight": "500",
          }
        }
      >
        E-mail : john@example.com
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flex": 0.2,
        "padding": 40,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#132440",
            "borderRadius": 5,
            "flexDirection": "row",
            "justifyContent": "center",
            "minHeight": 50,
            "padding": 10,
            "shadowOffset": {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 0.4,
          },
          undefined,
        ]
      }
      testID="logout-button"
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 15,
            "fontWeight": "700",
            "marginLeft": 10,
          }
        }
      >
        Log out
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
